keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,194 @@ the remaining requests in that queue start getting faster service.  In
+design and implementation of borrowing, until this is done we have two
+main options:
+- cap the `width` at the concurrency units assigned to the priority level
+- reject requests for which we won't be able to allocate enough concurrency
+  units
+
+To avoid breaking some users, we will proceed with the first option (when
+computing the cap we should also report requests that we believe are too
+wide for a given priority level - it would allow operators to adjust configs).
+That said, to accommodate for the inaccuracy here we will introduce a concept
+of `additional latency` for a request.  This basically means that after the
+request finishes in a real world, we still don't mark it as finished in
+the virtual world for `additional latency`.
+Adjusting virtual time of a queue to do that is trivial.  The other thing
+to tweak is to ensure that the concurrency units will not get available
+for other requests for that time (because currently all actions are
+triggerred by starting or finishing some request).  We will maintain that
+possibility by wrapping the handler into another one that will be sleeping
+for `additional latence` after the request is processed.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  Now in order to start processing a request, it has to accumulate
+`<width>` units of concurrency.
+
+The important requirement to recast now is fairness.  As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.  This justifes the above
+definition of adjusting the cost of the request to now be equal to
+`<width> x <processing latency>` (instead of just `<processing latency>`).
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification to the current dispatching algorithm:
+- as soon as we choose the request to dispatch (i.e. the queue from which
+  the first request should be dispatched), we start accumulating concurrency
+  units until we accumulate `<width>` and only then dispatch the request.
+  In other words, if the chosen request has width `<width>` and there
+  are less then `<width>` available seats, we don't dispatch any other request
+  (at a given priority level) until we will have `<width>` available seats
+  at which point we dispatch this request.
+  Such approach (as opposed to dispatching individual concurrency units
+  independently one-by-one) allows us to not waste too many seats and avoid
```

Comment:
(personally I would probably word this sentence stronger or omit it all together, I don't think there's any other reasonable way to do it without getting vastly more complex--not asking for a change though)